### ISSUE TITLE
MIR sanity check improvements

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#73832]
This issue collect various proposals for improvements to the MIR sanity check so that they do not get lost:

* Function calls and `DropAndReplace` could test that given and expected type match (https://github.com/rust-lang/rust/pull/72796#discussion_r446527001)
* `SetDiscriminant` could have some invariants checked (https://github.com/rust-lang/rust/pull/72796#discussion_r446624972)

**Comment by RalfJung** [comment]
@oli-obk re: `SetDiscriminant`, allowing that only for enums seems curiously asymmetric with `GetDiscriminant` which works for all ADTs.

**Comment by oli-obk** [comment]
Right, so let's check that the local is an `Adt` and that the discriminant id is actually a valid id for that type.

**Comment by RalfJung** [comment]
Does `GetDiscriminant` only work for ADTs or for all types? I forgot.

**Comment by oli-obk** [comment]
It works for all types (returning 0 for things that are neither Adt nor generator). That doesn't mean MIR building ever creates it and I don't think I've seen it before.

**Comment by ecstatic-morse** [comment]
`Rvalue::Discriminant` is also used for `TyKind::Generator`s.

**Comment by RalfJung** [comment]
> It works for all types (returning 0 for things that are neither Adt nor generator). That doesn't mean MIR building ever creates it and I don't think I've seen it before.

MIR building creates it for every `match` I think... or was it just every `match` of an ADT?
